DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/README.md

```diff
@@ -0,0 +1,238 @@
+   *  The word "LD" denotes a load and the word "ST" denotes the store. Feel free to explore the graphs. 
+2. In the Kernel pane, click on the label ```KernelPrefetch.B1``` to see a blue LD icon. 
+   *  Hovering over this icon should indicate that a prefetching LSU was inferred. 
+3. In the Kernel pane, click on the label ```KernelBurst.B1``` to see a blue LD icon. 
+   *  Hovering over this icon should indicate that a burst-coalesced LSU was inferred. 
+4. In the Kernel pane, click on the label ```KernelDefault.B1```to see a blue LD icon
+   *  Hovering over this icon, which LSU style was inferred?
+
+This view provides additional information about the LSU architecture selected in the details pane.
+
+### Area Analysis
+To check area usage:
+
+1. Navigate to Area Analysis of System (Area Analysis > Area Analysis of System). 
+   *  In this view you can see how much of the FPGA resources are being utilized (ALUTs, FFs, RAMs, MLABs, DSPs)
+   *  A design that takes less area will use less of these resources   
+2. In the center pane, expand the heading of Kernel System
+   *  Three kernels should be listed: ```KernelBurst```, ```KernelDefault```, ```KernelPrefetch```
+   *  Examine which kernel uses the most resources. 
+
+The ```KernelPrefetch``` should use less FPGA resources, as the design is a memory access pattern that lends itself to that LSU style best. 
+
+### Trade-offs
+
+There are many different ways to configure an LSU. As a programmer, the implementation you should choose depends on your design constraints.
+
+If your design is limited by the available FPGA resources, you can try the following:
+  - Request a prefetching LSU instead of a burst-coalesed LSU for loads.
+  - Disable the prefetching LSU, the burst-coalesced LSU, and caching, to infer a pipelined LSU. 
+  - Disable the LSU cache modifier, to prevent the compiler inferring any caching, especially if a cache will not be helpful for your design
+
+However, configuring an LSU solely based on area may compromise throughput. In this tutorial the access pattern is ideal for the prefetch LSU and it will acheive the same throughput as the burst coalesced, but this is not always the case. 
+  
+For more details on the descriptions of LSU controls, styles, and modifiers please refer the to the LSU Controls section in the [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide). 
+
+
+## Running the Sample
+
+ 1. Run the sample on the FPGA emulator (the kernel executes on the CPU):
+     ```
+     ./lsu_control.fpga_emu     (Linux)
+     lsu_control.fpga_emu.exe   (Windows)
+     ```
+2. Run the sample on the FPGA device:
+     ```
+     ./lsu_control.fpga         (Linux)
+     ```
+
+### Example of Output
+
+```
+Kernel throughput with prefetch LSU: 1040.11 MB/s 
+Kernel throughput with burst-coalesced LSU: 1035.82 MB/s 
+Kernel throughput without LSU controls: 1040.61 MB/s
```

Comment:
Do you have more motivating examples?

Reply:
Can you clarify what you mean?

Reply:
That *without* using LSU control you get *better* performance!!! :-)
It would be better to have an example showing that the FPGA extensions are useful for FPGA programming, if I understand correctly this tutorial.

Reply:
TLDR; I agree this is lack-luster but at this point we don't have a more motivating example.

@keryell the reason that _without_ attributes performs better is that the compiler chooses the best LSU using heuristics which, in this case, are good (it is likely creating an aligned prefetching LSU).

The point of this tutorial is to show users how to use the attributes. I agree that this is a bit lack-luster and there may be an argument that this doesn't do much more than documentation. It would be nice to find an example where using the attributes provides a win, especially compared to the default compiler behaviour. However, there are complications with putting that in a tutorial; if the compiler heuristics change, the tutorial may break.

Reply:
@keryell, I completely agree with @tyoungsc. 

Features LSU control exist so that you can override the compiler's heuristics if you disagree with the choices made by default. This can be to improve performance (if the compiler simply picked wrong) or to alter the area/performance tradeoff. The heuristics are generally quite good. While it's probably possible to find a fairly simple piece of code that outsmarts the heuristic, we recommend against devising tutorials around such examples. Invariably the tutorial "breaks" a few releases better when the compiler improves.